Allow braille printing (cups-brf) in selinux

zdohnal opened the pull request:
cups-brf gets data from getpwnam() to switch itself to the user who
issued the print job and then creates BRF dir and output files in user
home directories (/home/<username>/BRF for regular user, /root/BRF for
root).

sthibaul commented:
(this is very much like cups-pdf)

zpytela commented on policy/modules/contrib/cups.te:
```diff
@@ -594,6 +598,27 @@ optional_policy(`
+#
+
+# when printing as a root - to create /root/BRF/<file>
+allow cups_brf_t self:capability dac_override;
```
Can you share some details? What happens on the filesystem when printing as root?

zdohnal replied:
cups-brf backend calls getpwnam() to get info about the user who started print job and then the backend sets UID and GID to match with the issuer.
Then the backend uses umask() to 'act as the user' by forbidding permissions for 'group' and 'other' classes and creates BRF directory in user home directory (```/root``` for root user), creates a file in it, write filters' output in it and closes it.

To be honest, this is my first time writing a selinux policy, so I followed the suggestions from ```audit2allow``` command after I'd got a selinux denial, so I was updating the policy until the print job passed. Do feel free to let me know if I should approach this differently.

zpytela replied:
Please also update the commit message with the information for future reference.

zdohnal replied:
The commit message was updated to:

```
   Allow braille printing in selinux

    Description of the process:

    cups-brf backend calls getpwnam() to get info about the user who started
    print job and then the backend sets its UID and GID to match with the
    issuer.
    After that the backend uses umask() to 'act as the user' by forbidding
    permissions for 'group' and 'other' classes and creates BRF directory in
    user home directory (/root for root user, /home/<username> for others),
    creates a file in the directory, write filters' output in the file
    and closes it.
```

zpytela commented on policy/modules/contrib/cups.te:
```diff
@@ -594,6 +598,27 @@ optional_policy(`
+# sets its uid/gid to be run under user
+auth_read_passwd(cups_brf_t)
+init_search_pid_dirs(cups_brf_t)
+sssd_search_lib(cups_brf_t)
```
sssd is a module, its interfaces calls should be in an optional block.
authlogin, init, userdomain actually are modules, too, but they hardly can be removed from the system, so it's better to leave them where they are.

zdohnal replied:
@zpytela do you mean like this:

```
optional_policy(`
  sssd_search_lib(cups_brf_t)
')
```

?

zpytela replied:
Yes, just move it down (L621 or so).

zdohnal replied:
Moved down.

zdohnal commented:
@zpytela ping questions in review

zpytela commented:
> @zpytela ping questions in review

I am sorry, I completely missed the notifications.

zdohnal committed:
Allow braille printing in selinux

Description of the process:

cups-brf backend calls getpwnam() to get info about the user who started
print job and then the backend sets its UID and GID to match with the
issuer.
After that the backend uses umask() to 'act as the user' by forbidding
permissions for 'group' and 'other' classes and creates BRF directory in
user home directory (/root for root user, /home/<username> for others),
creates a file in the directory, write filters' output in the file
and closes it.

zdohnal commented:
@zpytela updated, would you mind looking into the current PR?

zpytela commented:
The commit well matches cups-pdf so as such it is fine. However, there is one troublesome thing: with `userdom_manage_user_home_content_dirs(cups_brf_t)` we are giving to the plugin unnecessary permissions to manage user's home.

Will it be possible to change the software behaviour to use some XDG adviced directory, e. g. ~/.config/cups/brf?

zdohnal commented:
> Will it be possible to change the software behaviour to use some XDG adviced directory, e. g. ~/.config/cups/brf?

@zpytela  The file in the created BRF dir is a printout of a document - IMHO it doesn't make much sense to put it into a dir which is meant for configurations...

Is it possible to somehow define in selinux policy that the backend can only create and access only BRF directory in home dir, and then open/write/close files in BRF dir?

zpytela commented:
> > Will it be possible to change the software behaviour to use some XDG adviced directory, e. g. ~/.config/cups/brf?
> 
> @zpytela The file in the created BRF dir is a printout of a document - IMHO it doesn't make much sense to put it into a dir which is meant for configurations...

What I meant was ~/.cache, but there are other "well-known" directories, too.

> 
> Is it possible to somehow define in selinux policy that the backend can only create and access only BRF directory in home dir, and then open/write/close files in BRF dir?

Policy can give access to a particular directory, but if it does not exist, permissions to modify user's home dir are required if it is in user's home dir. That's why a dedicated data/cache/runtime directory is a better choice.

https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

zdohnal commented:
@sthibaul Hi Sam,

as cups-brf author, would you mind looking into the proposal @zpytela mentioned? Is it acceptable to change the dir?

sthibaul commented:
> Is it acceptable to change the dir?

Well, as much as it would be for the cups-pdf filter. The problem is that if the file is stuffed into a hidden directory, there is very little probability that the user will find it...

zdohnal commented:
@zpytela ,

based on @sthibaul review, the changing of dir would reduce the output visibility, so cups-brf cannot use a XDG dir for output.

zpytela commented:
I see. This needs to be discussed in our team, I'll let you know.

zdohnal commented:
Hi @zpytela ,

any news regarding the PR?

zpytela commented:
It is still considered unfortunate and we will appreciate using a standardized directory in the future in order not to allow arbitrary domains manage user's home. Merging though as the risk has been assessed minimal. Thank you for your contribution.
